### DISCUSSION TITLE
Getting Started - and other learning resources

### DISCUSSION THREAD

**Discussion by juanmaguitar** [WordPress/gutenberg#52894] [Interactivity API]
Ready to add some interactivity to your blocks? Although it is still experimental, you can begin to explore the Interactivity API potential and provide valuable feedback. Remember, your feedback and suggestions will help us shape the API and improve the tool for everyone. So let's get started! 🙂🚀

- If you have problems or suggestions following the Getting Started guide, you can comment on this discussion to let us know or open a pull request to improve the guide. 
- If you want to share your feedback or have questions about the Interactivity API, please open a new discussion in the [Interactivity API category](https://github.com/WordPress/gutenberg/discussions/categories/interactivity-api). You can also use the WordPress Slack's [#core-interactivity](https://wordpress.slack.com/archives/C071CRKGKUP) channel.

### Resources
- ⭐️ **[Getting Started guide](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/iapi-quick-start-guide/)**: Please follow the Getting Started guide to learn how to scaffold a new project and create your first interactive blocks.
- ⭐️ **[API Reference](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/api-reference/)**: Check this page for detailed explanations and examples of the directives and the store.
- ⭐ **[A first look at the Interactivity API](https://developer.wordpress.org/news/2024/04/11/a-first-look-at-the-interactivity-api/)**: A tutorial to get started with the basics of the Interactivity API.
&nbsp;
- **[Dev note: The Interactivity API in WP 6.5](https://make.wordpress.org/core/2024/03/04/interactivity-api-dev-note/)** – First public version of the Interactivity API in WordPress Core | make.wordpress.org/core blog post
- **[Proposal: The Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/)** – The initial proposal for a better developer experience in building interactive blocks | make.wordpress.org/core blog post
- **Developer Hours sessions**: [First session](https://www.youtube.com/watch?v=RXNoyP2ZiS8&t=664s) and [second session](https://www.youtube.com/watch?v=6ghbrhyAcvA)
- [Building interactive blocks: a step-by-step workshop](https://wordpress.tv/2023/10/01/building-interactive-blocks-a-step-by-step-workshop/) | WCEU Workshop
- Core blocks using the Interactivity API: [Navigation block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/navigation), [Image block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/image), [Search block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/search), [File block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/file), [Query block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/query).
- [New WordPress Developer APIs – The Interactivity API](https://wordpress.tv/2023/12/15/new-wordpress-developer-apis-the-interactivity-api/) | learn.wordpress.org session by @jonathanbossenger (see [slides](https://www.slideshare.net/psykro/new-wordpress-developer-apis-the-interactivity-api))

### Examples/ Demos

- **WP Movies Demo**: [wpmovies.dev](http://wpmovies.dev/) and its [wp-movies-demo](https://github.com/WordPress/wp-movies-demo) repository
- Examples using the Interactivity API at [block-development-examples](https://github.com/WordPress/block-development-examples):
  - [`interactivity-api-block-833d15`](https://github.com/WordPress/block-development-examples/tree/trunk/plugins/833d15)
  - [`interactivity-api-countdown-3cd73e`](https://github.com/WordPress/block-development-examples/tree/trunk/plugins/interactivity-api-countdown-3cd73e)
  - [`interactivity-api-quiz-1835fa`](https://github.com/WordPress/block-development-examples/tree/trunk/plugins/interactivity-api-quiz-1835fa)
---

**Documentation is still in progress**. In the meantime, please do not hesitate to start a new discussion for any questions you may have regarding how to use it, the different APIs, or recommended patterns for specific scenarios. If you are willing to help with the documentation, [please say so in this discussion](https://github.com/WordPress/gutenberg/discussions/51928), and we'll coordinate everyone's efforts.

**Comment by FawadNL** [comment]
When will the WP movies theme and its blocks get updated to use the latest interactivity api?

**Reply by luisherranz** [reply]
As soon as we ship the ability to do _full page_ client-side navigation in Gutenberg.

**Reply by FawadNL** [reply]
Thanks for the prompt reply. Any ETA for the full client side navigation?

**Reply by luisherranz** [reply]
It's one of the things we would like to focus on [after the 1.0 public release](https://github.com/WordPress/gutenberg/issues/56803).

**Reply by ts-mz** [reply]
Hi. Sorry for jumping in here - I'm in the process of transitioning a simple wix site to a WP block theme and the wix site is using fade transitions for the content area. So I had the idea to check if this would be possible using the interactivity API and thus doing something useful for myself by learning about using interactivity during this project (because I remembered instant transitions from the wpmovies demo). But reading this I'm getting the impression that something like content area transitions is not yet really supported but was handcrafted into the wpmovies demo? Is that impression correct? Thanks!

**Reply by luisherranz** [reply]
May I ask what you mean by fade transitions? Is it when the browser doesn't need to refresh the page (like in wpmovies.dev)? Or something more specific?

**Reply by ts-mz** [reply]
Thanks for your reply! And sorry for being unclear. Yes, initially the first, content loading without page loading. In the long run, I'd hope to be able to use the interactivity api to achieve actual fading page transitions, similar to this - https://reactcommunity.org/react-transition-group/ (unless there's a better approach by then).

**Reply by luisherranz** [reply]
Ok, thanks!

In WP 6.5 we're releasing a type of navigation that we call _region-based client-side navigation_. Instead of replacing the content of the full page, it replaces the content inside some regions. You can see it in action in WP 6.4 in the Query block (disable the "Force full page reloads" setting). The reason to release the _region-based_ navigation first is that it is a bit simpler and it's compatible with more sites. We'll prepare a guide in the docs to explain how it works and how people can leverage the new `core/router` store to use this type of navigation (please be aware that the guide might not be ready by the release of WP 6.5 and come later).

The _full-page_ client-side navigation will take a bit more. We want to release it this year, so maybe in WP 6.7, but I can't make any promises as it has challenges that are not so easy to solve.

We also want to work on adding support for the native [View Transitions API](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API) for both the _region-based_ and _full-page_ client-side navigation. I think this might be what you're looking for. Again, this is something we'd like to work on later this year, but we can't make any promises as its design also presents some challenges.

**Reply by ts-mz** [reply]
Wow, thanks for the extensive reply! Looking forward to trying all this when it drops! Will have a loook at the query block =)

**Comment by juanmaguitar** [comment]
## Learning Resources from the Community

- [#77 – Mario Santos and Luis Herranz on What the Interactivity API Is and How You Can Use It](https://wptavern.com/podcast/77-mario-santos-and-luis-herranz-on-what-the-interactivity-api-is-and-how-you-can-use-it) | WP Tavern podcast
- [🧑‍🍳 Block Developer Cookbook Recipe: The Interactivity API 🧑‍🍳](https://www.youtube.com/watch?v=Sj61X_Xtfzc) by @ryanwelcher

**Comment by gaambo** [comment]
FYI: If anyone is experimenting with the interactivity API in custom blocks with the newest Gutenberg version(s), you have to change the `view` field in your `block.json` to `viewModule` and change the build task to `wp-scripts start --experimental-modules` - for that you need to use `@wordpress/scripts ^27.1.0` and `@wordpress/interactivity ^4.0.1`. It's kind of documented [in the @wordpress/scripts package](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-scripts/#build), but I had to fiddle around yesterday a little bit to get it working. 

If you're also extending the default webpack config of the scripts package: With `--experimental-modules` the defaultConfig is not just an object but an [array of two configs](https://github.com/WordPress/gutenberg/blob/89a15ccf9a99b77a04a3b5008a4f6d62c3784c7e/packages/scripts/config/webpack.config.js#L435). If you want to add other entrypoints, you probably want to add it to the first config object (being the "normal" one, the second one is the module config).

**Reply by luisherranz** [reply]
I've just published the migration guide for Gutenberg 17.6, which includes other breaking changes: 

- https://github.com/WordPress/gutenberg/discussions/52906#discussioncomment-8344138

**Comment by nextgenthemes** [comment]
From the merge announcement.

> Backward compatible. Ensuring compatibility with both classic and block themes and optionally with other JavaScript frameworks, though it’s advised to use the API as the primary method. It also works with hooks and internationalization.

But every example is with blocks, so is it? Can I use this in classic Themes/Plugins without Blocks? I just ported my setting page to Alpine.js (for no real reason) and now I learn about this, and I am exited and annoyed at the same time, as I obviously would like to port it again to this as it looks very similar and was inspired by it right?

I need to support classic editor, I am not even sure if it's possible to use blocks for the admin screens, I guess not, but I do not want to anyway. Additionally, I would love to use it on plugins generated HTML that does not use blocks.

Would be happy for any clarification and a basic tutorial on how to use it without Blocks.

**Reply by luisherranz** [reply]
Yes, it is possible to use it in classic themes as well 🙂

You can take a look at this TwentyTwentyOne PR:
- https://github.com/WordPress/wordpress-develop/pull/5795

And the documentation of [`wp_interactivity_process_directives`](https://developer.wordpress.org/reference/functions/wp_interactivity_process_directives/) to process the directives present outside of blocks.

Let us know how it goes!
